drivers/VirtualWatchdog.h

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2019 Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_VIRTUAL_WATCHDOG_H
+#define MBED_VIRTUAL_WATCHDOG_H
+
+#ifdef DEVICE_WATCHDOG
+
+#include <cstdio>
+#include "platform/mbed_error.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_power_mgmt.h"
+#include "platform/mbed_assert.h"
+#include "platform/SingletonPtr.h"
+
+#if DEVICE_LPTICKER
+#include "drivers/LowPowerTicker.h"
+#else
+#include "drivers/Ticker.h"
+#endif
+namespace mbed {
+
+/** \addtogroup drivers */
+
+/** VirtualWatchdog should be used for applications where you use multiple services requiring watchdog functionality.
+ *   Use `Watchdog` driver for simple uses cases like bootloader.
+ *
+ *  A system timer that will reset the system in the case of system failures or
+ *  malfunctions.
+ *
+ * Example:
+ * @code
+ *
+ * Example:
+ * @code
+ *
+ * VirtualWatchdog virtual_watchdog(300,"Software Watchdog");
+ * virtual_watchdog.start();
+ *
+ * while (true) {
+ *    virtual_watchdog.kick();
+ *
+ *    // Application code
+ * }
+ * @endcode
+ * @ingroup platform
+ */
+class VirtualWatchdog {
+public:
+
+    /** Constructor configured with timeout and name for this software watchdog instance.
+     *
+     * We start hardware watchdog on  the first run. This set-up uses default watchdog timeout set via config value
+     * MBED_CONF_TARGET_WATCHDOG_TIMEOUT.
+     *
+     */
+    VirtualWatchdog(uint32_t timeout = 1000, const char *const str = NULL);
+    ~VirtualWatchdog();
+public:
+
+    /** Start an independent watchdog timer with specified parameters.
+     *
+     * Assert for multiple calls of start.
+     */
+    void start();
+
+    /** This stops the watchdog timer.
+     *
+     * Calling this function disables any running
+     * watchdog timers if the platform supports them.
+     *
+     * Assert without calling start.
+     */
+    void stop();
```

Comment:
Should we return bool?

Reply:
stop asserts if not initialized - the only error that can happen (catched via asserts in start/stop duo). When we would return false?

Reply:
When you can't stop the watchdog:
start -> starts the HW
start -> adds to the list
stop -> ok, removes from the list
stop -> fail as this particular HW doesn't support stopping

Reply:
Ah I see we never actually stop the HW timer.